server/server.go

```diff
@@ -0,0 +1,130 @@
+package main
+
+import (
+	"errors"
+	"io"
+	"log"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+type server struct {
+	db  *database
+	mux *http.ServeMux
+}
+
+func (s *server) routes() {
+	s.mux.HandleFunc("/db", s.metricsMiddleware(s.handleDB()))
+	s.registerMetrics()
+
+	s.mux.HandleFunc("/*", s.handleBadPath())
+}
+
+func (s *server) handleDB() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		key := r.URL.Query().Get("key")
+		switch r.Method {
+		case http.MethodDelete:
+			s.handleDelete(w, key)
+		case http.MethodGet:
+			s.handleGet(w, key)
+		case http.MethodPut:
+			s.handlePut(w, r, key)
+		default:
+			s.handleNotImplemented(w)
+		}
+	}
+}
+
+func (s *server) handleDelete(w http.ResponseWriter, key string) {
+	_, ok := s.db.get(key)
+	if !ok {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	s.db.delete(key)
+	w.WriteHeader(http.StatusOK)
+}
+
+func (s *server) handleGet(w http.ResponseWriter, key string) {
+	value, ok := s.db.get(key)
+	if !ok {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	_, err := w.Write([]byte(value))
+	if err != nil {
+		log.Println("Error writing response:", err)
+		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+	}
+}
+
+// according to rfc guidelines PUT should create or replace resources
+// https://www.rfc-editor.org/rfc/rfc2616#section-9.6
+func (s *server) handlePut(w http.ResponseWriter, r *http.Request, key string) {
+	_, ok := s.db.get(key)
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		log.Println("Error reading body:", err)
+		http.Error(w, "Error reading body", http.StatusBadRequest)
+		return
+	}
+	err = s.db.put(key, string(body))
```

Comment:
This is a bit unlikely, but still possible: the mutex only protects access inside of its methods so another request could make changes to the database between `db.get` and `db.put`. From what I remember other solutions did not have this problem